File: jawanndenn/markup.py
# Copyright (C) 2016 Sebastian Pipping <sebastian@pipping.org>
# Licensed under GNU Affero GPL v3 or later

_REPLACEMENTS_IN_ORDER = (
    ("**", "<strong>", "</strong>"),
    ("*", "<em>", "</em>"),
    ("__", "<strong>", "</strong>"),
    ("_", "<em>", "</em>"),
    ("`", "<tt>", "</tt>"),
)

_CLOSING_OF = {prefix: closing for prefix, _, closing in _REPLACEMENTS_IN_ORDER}


def safe_html(text):
    if not isinstance(text, str):
        raise ValueError("Not a string: %s" % text)

    # KEEP IN SYNC with javascript client side!

    text = text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")

    chunks = []

    opened = []
    while text:
        for prefix, opening, closing in _REPLACEMENTS_IN_ORDER:
            if text.startswith(prefix):
                if opened and opened[-1] == prefix:
                    # Close tag
                    chunks.append(closing)
                    opened.pop()
                else:
                    # Open tag
                    chunks.append(opening)
                    opened.append(prefix)

                text = text[len(prefix) :]
                break
        else:
            chunks.append(text[0])
            text = text[1:]

    # Close all unclosed tags
    for prefix in reversed(opened):
        chunks.append(_CLOSING_OF[prefix])

    return "".join(chunks)

File: jawanndenn/frontend/src/markup.ts
// Copyright (c) 2025 Sebastian Pipping <sebastian@pipping.org>
// Licensed under GNU Affero GPL v3 or later

const REPLACEMENTS_IN_ORDER = [
  ['**', '<strong>', '</strong>'],
  ['*', '<em>', '</em>'],
  ['__', '<strong>', '</strong>'],
  ['_', '<em>', '</em>'],
  ['`', '<tt>', '</tt>'],
];

const CLOSING_OF = {};
REPLACEMENTS_IN_ORDER.forEach((row) => {
  const prefix = row[0];
  const closing = row[2];
  CLOSING_OF[prefix] = closing;
});

// Excapes HTML and renders subset of markdown
const textToSafeHtml = (text: string): string => {
  // KEEP IN SYNC with python server side!
  text = text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');

  const chunks = [];

  const opened = [];
  while (text.length) {
    let matched = false;

    REPLACEMENTS_IN_ORDER.forEach((row) => {
      const prefix = row[0];
      const opening = row[1];
      const closing = row[2];

      if (text.startsWith(prefix)) {
        if (opened.length && opened[opened.length - 1] == prefix) {
          // Close tag
          chunks.push(closing);
          opened.pop();
        } else {
          // Open tag
          chunks.push(opening);
          opened.push(prefix);
        }

        text = text.slice(prefix.length);

        matched = true;
        return false;
      }
    });

    if (!matched) {
      chunks.push(text[0]);
      text = text.slice(1);
    }
  }

  // Close all unclosed tags
  opened.reverse();
  opened.forEach((prefix) => {
    chunks.push(CLOSING_OF[prefix]);
  });

  return chunks.join('');
};

export { textToSafeHtml };
